src/environment.py

```diff
@@ -0,0 +1,298 @@
+        "MM_EMAILSETTINGS_SENDPUSHNOTIFICATIONS": "true",
+        "MM_EMAILSETTINGS_PUSHNOTIFICATIONCONTENTS": contents,
+        "MM_EMAILSETTINGS_PUSHNOTIFICATIONSERVER": config.get("push_notification_server"),
+    }
+
+
+def _env_for_sso(config: dict, site_url: str) -> dict:
+    """Return settings to use login.ubuntu.com via SAML for single sign-on.
+
+    SAML_IDP_CRT must be generated and installed manually by a human (see README.md).
+
+    Args:
+        config: dict of the charm's configuration
+        site_url: public facing URL of mattermost
+
+    Returns:
+        dict of settings to use login.ubuntu.com via SAML for single sign-on.
+    """
+    if not config.get("sso") or any(not config.get(setting) for setting in REQUIRED_SSO_SETTINGS):
+        return {}
+    site_hostname = urlparse(site_url).hostname
+    use_experimental_saml_library = (
+        "true" if config.get("use_experimental_saml_library") else "false"
+    )
+
+    return {
+        "MM_EMAILSETTINGS_ENABLESIGNINWITHEMAIL": "false",
+        "MM_EMAILSETTINGS_ENABLESIGNINWITHUSERNAME": "false",
+        "MM_EMAILSETTINGS_ENABLESIGNUPWITHEMAIL": "false",
+        "MM_SAMLSETTINGS_ENABLE": "true",
+        "MM_SAMLSETTINGS_IDPURL": "https://login.ubuntu.com/saml/",
+        "MM_SAMLSETTINGS_VERIFY": "true",
+        "MM_SAMLSETTINGS_ENCRYPT": "false",  # per POC
+        "MM_SAMLSETTINGS_IDPDESCRIPTORURL": "https://login.ubuntu.com",
+        "MM_SAMLSETTINGS_SERVICEPROVIDERIDENTIFIER": "https://login.ubuntu.com",
+        "MM_SAMLSETTINGS_IDPMETADATAURL": "https://login.ubuntu.com/+saml/metadata",
+        "MM_SAMLSETTINGS_ASSERTIONCONSUMERSERVICEURL": f"https://{site_hostname}/login/sso/saml",
+        "MM_SAMLSETTINGS_LOGINBUTTONTEXT": "Ubuntu SSO",
+        "MM_SAMLSETTINGS_EMAILATTRIBUTE": "email",
+        "MM_SAMLSETTINGS_USERNAMEATTRIBUTE": "username",
+        "MM_SAMLSETTINGS_IDATTRIBUTE": "openid",
+        "MM_SAMLSETTINGS_FIRSTNAMEATTRIBUTE": "fullname",
+        "MM_SAMLSETTINGS_LASTNAMEATTRIBUTE": "",
+        "MM_SAMLSETTINGS_IDPCERTIFICATEFILE": SAML_IDP_CRT,
+        # If not set, we have to install xmlsec1, and Mattermost forks on every login(!).
+        "MM_EXPERIMENTALSETTINGS_USENEWSAMLLIBRARY": use_experimental_saml_library,
+    }
+
+
+def _env_for_smtp(config: dict) -> dict:
+    """Return settings for an outgoing SMTP relay.
+
+    Args:
+        config: dict of the charm's configuration
+
+    Returns:
+        dict of settings for an outgoing SMTP relay.
+    """
+    if any(
+        (
+            setting not in config
+            for setting in (
+                "smtp_host",
+                "smtp_connection_security",
+                "smtp_from_address",
+                "smtp_reply_to_address",
+                "smtp_password",
+                "smtp_user",
+                "smtp_port",
+            )
+        )
+    ):
+        return {}
+
+    if not config["smtp_host"]:
+        return {}
+
+    enable_smtp_auth = "false"
```

Comment:
what about `enable_smtp_auth = config["smtp_user"] and config["smtp_password"]`?

Reply:
I can't do that since `enable_smtp_auth` is a `str`

Reply:
what about `enable_smtp_auth = str(config["smtp_user"] and config["smtp_password"]).lower()`?

Reply:
I used Yanks' proposal